src/test/fuzz/crypto_diff_fuzz_chacha20.cpp

```diff
@@ -0,0 +1,334 @@
+
+        x0 = XOR(x0, U8TO32_LITTLE(m + 0));
+        x1 = XOR(x1, U8TO32_LITTLE(m + 4));
+        x2 = XOR(x2, U8TO32_LITTLE(m + 8));
+        x3 = XOR(x3, U8TO32_LITTLE(m + 12));
+        x4 = XOR(x4, U8TO32_LITTLE(m + 16));
+        x5 = XOR(x5, U8TO32_LITTLE(m + 20));
+        x6 = XOR(x6, U8TO32_LITTLE(m + 24));
+        x7 = XOR(x7, U8TO32_LITTLE(m + 28));
+        x8 = XOR(x8, U8TO32_LITTLE(m + 32));
+        x9 = XOR(x9, U8TO32_LITTLE(m + 36));
+        x10 = XOR(x10, U8TO32_LITTLE(m + 40));
+        x11 = XOR(x11, U8TO32_LITTLE(m + 44));
+        x12 = XOR(x12, U8TO32_LITTLE(m + 48));
+        x13 = XOR(x13, U8TO32_LITTLE(m + 52));
+        x14 = XOR(x14, U8TO32_LITTLE(m + 56));
+        x15 = XOR(x15, U8TO32_LITTLE(m + 60));
+
+        j12 = PLUSONE(j12);
+        if (!j12) {
+            j13 = PLUSONE(j13);
+            /* stopping at 2^70 bytes per nonce is user's responsibility */
+        }
+
+        U32TO8_LITTLE(c + 0, x0);
+        U32TO8_LITTLE(c + 4, x1);
+        U32TO8_LITTLE(c + 8, x2);
+        U32TO8_LITTLE(c + 12, x3);
+        U32TO8_LITTLE(c + 16, x4);
+        U32TO8_LITTLE(c + 20, x5);
+        U32TO8_LITTLE(c + 24, x6);
+        U32TO8_LITTLE(c + 28, x7);
+        U32TO8_LITTLE(c + 32, x8);
+        U32TO8_LITTLE(c + 36, x9);
+        U32TO8_LITTLE(c + 40, x10);
+        U32TO8_LITTLE(c + 44, x11);
+        U32TO8_LITTLE(c + 48, x12);
+        U32TO8_LITTLE(c + 52, x13);
+        U32TO8_LITTLE(c + 56, x14);
+        U32TO8_LITTLE(c + 60, x15);
+
+        if (bytes <= 64) {
+            if (bytes < 64) {
+                for (i = 0; i < bytes; ++i)
+                    ctarget[i] = c[i];
+            }
+            x->input[12] = j12;
+            x->input[13] = j13;
+            return;
+        }
+        bytes -= 64;
+        c += 64;
+        m += 64;
+    }
+}
+
+void ECRYPT_keystream_bytes(ECRYPT_ctx* x, u8* stream, u32 bytes)
+{
+    u32 i;
+    for (i = 0; i < bytes; ++i)
+        stream[i] = 0;
+    ECRYPT_encrypt_bytes(x, stream, stream, bytes);
+}
+
+FUZZ_TARGET(crypto_diff_fuzz_chacha20)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+
+    ChaCha20 chacha20;
+    ECRYPT_ctx ctx;
+    // D. J. Bernstein doesn't initialise ctx to 0 while Bitcoin Core initialises chacha20 to 0 in the constructor
+    for (int i = 0; i < 16; i++) {
+        ctx.input[i] = 0;
+    }
+
+    if (fuzzed_data_provider.ConsumeBool()) {
+        const std::vector<unsigned char> key = ConsumeFixedLengthByteVector(fuzzed_data_provider, fuzzed_data_provider.ConsumeIntegralInRange<size_t>(16, 32));
+        chacha20 = ChaCha20{key.data(), key.size()};
+        ECRYPT_keysetup(&ctx, key.data(), key.size() * 8, 0);
+        // ECRYPT_keysetup() doesn't set the counter and nonce to 0 while SetKey() does
+        uint8_t iv[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+        ECRYPT_ivsetup(&ctx, iv);
+    }
+
+    while (fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(0, 100) >= 10) {
```

Comment:
how is this different from

```suggestion
    while (fuzzed_data_provider.ConsumeBool()) {
```

Also, could use `LIMITED_WHILE` to avoid unlimited runtime.

Reply:
> how is this different from

Increased the probability of entering the while loop as `ConsumeBool()` has only 50% chance of returning true while `ConsumeIntegralInRange<uint32_t>(0,100) >= 10` will have 90% chance of returning true.
Essentially we’re trying to get the test running longer for each fuzz seed so multiple functions are exercised per seed, especially since `ChaCha20` maintains an internal state for each seed.

Reply:
> Also, could use `LIMITED_WHILE` to avoid unlimited runtime.

Great suggestion! Added `LIMITED_WHILE`.

Reply:
> Essentially we’re trying to get the test running longer

Do you have data to support that claim? Fuzzing engines will store inputs that increase coverage data and once the data is stored, for replay it doesn't matter whether it was `ConsumeBool` or `ConsumeIntegralInRange` that evaluated to true.

Reply:
> Do you have data to support that claim?

Backing the claim with data was indeed a great suggestion, so we generated the data and plots to verify it.
Now we have results to dismiss that claim :)
Sharing our findings here:

<img src="https://cdn.discordapp.com/attachments/863100385152335883/878387475096170607/unknown.png"
	title="Plot" width="500" height="300" />
`x-axis denotes no of iterations`
`y-axis denotes frequency of each iteration`